src/viam/sessions_client.py

```diff
@@ -79,77 +89,78 @@ async def _recv_trailers(self, event: RecvTrailingMetadata):
-            if self._supported is False:
+        request = StartSessionRequest(resume=self._current_id)
+        try:
+            response = await self.client.StartSession(request)
+        except GRPCError as error:
+            if error.status == Status.UNIMPLEMENTED:
+                with self._lock:
+                    self._supported = _SupportedState.FALSE
                 return self._metadata
-
-            request = StartSessionRequest(resume=self._current_id)
-            response: Optional[StartSessionResponse] = None
-
-            try:
-                response = await self.client.StartSession(request)
-            except GRPCError as error:
-                if error.status == Status.UNIMPLEMENTED:
-                    self._supported = False
-                    return self._metadata
-                else:
-                    raise
             else:
-                if response is None:
-                    raise GRPCError(status=Status.INTERNAL, message="Expected response to start session")
+                raise
 
-                if response.heartbeat_window is None:
-                    raise GRPCError(status=Status.INTERNAL, message="Expected heartbeat window in response to start session")
+        if response is None:
+            raise GRPCError(status=Status.INTERNAL, message="Expected response to start session")
 
-                self._supported = True
-                self._heartbeat_interval = response.heartbeat_window.ToTimedelta()
-                self._current_id = response.id
+        if response.heartbeat_window is None:
+            raise GRPCError(status=Status.INTERNAL, message="Expected heartbeat window in response to start session")
 
-        # tick once to ensure heartbeats are supported
-        await self._heartbeat_tick()
+        with self._lock:
+            self._supported = _SupportedState.TRUE
+            self._heartbeat_interval = response.heartbeat_window.ToTimedelta()
+            self._current_id = response.id
 
-        if self._supported:
-            # We send heartbeats slightly faster than the interval window to
-            # ensure that we don't fall outside of it and expire the session.
-            wait = self._heartbeat_interval.total_seconds() / 5
-            asyncio.create_task(self._heartbeat_task(wait), name=f"{_TASK_PREFIX}-heartbeat")
+        # tick once to ensure heartbeats are supported
+        await self._heartbeat_tick(self.client)
+
+        with self._lock:
+            if self._supported == _SupportedState.TRUE:
+                # We send heartbeats slightly faster than the interval window to
+                # ensure that we don't fall outside of it and expire the session.
+                wait = self._heartbeat_interval.total_seconds() / 5
+
+                self._thread = Thread(
+                    name="heartbeat-thread",
+                    target=asyncio.run,
+                    args=(self._heartbeat_process(wait),),
+                    daemon=True,
```

Comment:
Are we sure this should be daemonized?

Reply:
main issue I ran into a non-daemonized thread is that it doesn't stop on a ctrl+c, and since the daemonic thread just dies on program exit, the cleanup can be handled like normal (since only one rust connection is created).